Fix tests to ignore the startup file

rsrock opened the pull request:
To find the offending tests, it was helpful to use
a .juliarc.jl with just `error("don't include me")`.

Fixes #17642

rsrock committed:
Fix tests to ignore the startup file

To find the offending tests, it was helpful to use
a .juliarc.jl with just `error("don't include me")`.

rsrock commented on test/cmdlineargs.jl:
```diff
@@ -277,21 +270,33 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # warnings in ~/.juliarc.jl cause the tests to freeze
+    let JL_OPTIONS_STARTUPFILE_ON = 1,
+        JL_OPTIONS_STARTUPFILE_OFF = 2
+        # @test parse(Int,readchomp(`$exename -E "Base.JLOptions().startupfile" --startup-file=yes`)) == JL_OPTIONS_STARTUPFILE_ON
```
This is a bit of a problem.  I don't know how to properly test that the startup file is being used, if a broken startup file causes the tests to hang.

vtjnash replied:
Set the HOME env variable to a controlled location?

rsrock commented on test/cmdlineargs.jl:
```diff
@@ -69,6 +69,7 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
 
     # --procs
+    # FIXME: is the -p 2 option incompatible with --startup-file=no ?
     @test readchomp(`$exename -q -p 2 -e "println(nworkers())"`) == "2"
```
Here's a problem:

```❯ ./julia -q --startup-file=no -e "println(nworkers())"
1

rrock at Uluru in ~/Documents/Code/julia5  rsr/test-ignore-startup
❯ ./julia -q -p 2 --startup-file=no -e "println(nworkers())"
ERROR: LoadError: Don't include me!
 in error(::String) at ./error.jl:21
 in include_from_node1(::String) at ./loading.jl:426
 in try_include at ./client.jl:200 [inlined]
 in load_juliarc() at ./client.jl:278
 in process_options(::Base.JLOptions) at ./client.jl:216
 in _start() at ./client.jl:318
while loading /Users/rrock/.juliarc.jl, in expression starting on line 1
ERROR: LoadError: Don't include me!
 in error(::String) at ./error.jl:21
 in include_from_node1(::String) at ./loading.jl:426
 in try_include at ./client.jl:200 [inlined]
 in load_juliarc() at ./client.jl:278
 in process_options(::Base.JLOptions) at ./client.jl:216
 in _start() at ./client.jl:318
while loading /Users/rrock/.juliarc.jl, in expression starting on line 1

[hangs here]
```

It looks like the `-p 2` option ignores the `--startup-file=no` option.  Should I file a separate issue for this?

vtjnash replied:
Yes

tkelman commented on test/cmdlineargs.jl:
```diff
@@ -277,21 +270,33 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
 
+let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # --startup-file
+    # Note: cannot cleanly test for --startup-file=yes, because errors or deprication
```
deprecation

rsrock commented:
Ok, I addressed the comments and filed a separate issue for `julia -p`.  I have one remaining question about setting the HOME variable, because I doubt what I did here will work on Windows.

tkelman commented on test/cmdlineargs.jl:
```diff
@@ -277,21 +269,32 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    let JL_OPTIONS_STARTUPFILE_ON = 1,
+        JL_OPTIONS_STARTUPFILE_OFF = 2
+        # `HOME=/tmp` is to bypass any errors in user startup files, which cause the tests to hang
+        @test parse(Int,readchomp(`HOME=/tmp $exename -E "Base.JLOptions().startupfile" --startup-file=yes`)) == JL_OPTIONS_STARTUPFILE_ON
```
you'll probably need to use the `setenv` function for this

rsrock replied:
Aha, the julia `setenv` function.  That helps.

rsrock commented:
The Travis failure looks unrelated.  I think this part is ready to go.

tkelman commented on test/cmdlineargs.jl:
```diff
@@ -277,21 +269,32 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # --startup-file
+    let JL_OPTIONS_STARTUPFILE_ON = 1,
+        JL_OPTIONS_STARTUPFILE_OFF = 2
+        # `HOME=/tmp` to avoid errors in the user .juliarc.jl, which hangs the tests.  Issue #
```
add the actual issue number?

rsrock replied:
That would be good to have, wouldn't it...

vtjnash commented:
lgtm

rsrock commented:
The issue number is in (thanks!), so I think this is ready
